File: Components/Reporting/VB/ASPxReportDemos/App_Code/WebDocumentViewerDrillThroughProcessor.vb
Imports Microsoft.VisualBasic
Imports System.Web
Imports System.Collections.Generic
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Web.WebDocumentViewer
Imports XtraReportsDemos.PollingReport

Public Class WebDocumentViewerDrillThroughProcessor
	Implements IWebDocumentViewerDrillThroughProcessor
	Private Function GetResultingData() As IList(Of ResultingData)
		If HttpContext.Current.Session("resultingData") Is Nothing Then
			HttpContext.Current.Session("resultingData") = New List(Of ResultingData)()
		End If
		Return CType(HttpContext.Current.Session("resultingData"), List(Of ResultingData))
	End Function
	Private Function IncrementPollCount() As Integer
		Dim pollCount As Integer = 3
		If HttpContext.Current.Session("pollCount") IsNot Nothing Then
			pollCount = CInt(Fix(HttpContext.Current.Session("pollCount")))
		End If
		pollCount += 1
		HttpContext.Current.Session("pollCount") = pollCount
		Return pollCount
	End Function
	Private Function CreateReport(ByVal context As DrillThroughContext) As XtraReport Implements IWebDocumentViewerDrillThroughProcessor.CreateReport
		If context.CustomData = "showResults" Then
			context.AssignEditingFieldValuesToReport()
			Dim data As IList(Of ResultingData) = GetResultingData()
			CType(context.Report, PollingReport).UpdateData(data)
			Dim resultingReport = New ResultingReport()
			resultingReport.ResultingReportDS.DataSource = data
			resultingReport.Parameters("paramPollCount").Value = IncrementPollCount()
			Return resultingReport
		Else
			Return New PollingReport()
		End If
	End Function
End Class

File: Components/Reporting/CS/MVCxReportDemos/Code/WebDocumentViewerDrillThroughProcessor.cs
using System.Web;
using System.Collections.Generic;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Web.WebDocumentViewer;
using XtraReportsDemos.PollingReport;

public class WebDocumentViewerDrillThroughProcessor : IWebDocumentViewerDrillThroughProcessor {
    IList<ResultingData> GetResultingData() {
        if(HttpContext.Current.Session["resultingData"] == null) {
            HttpContext.Current.Session["resultingData"] = new List<ResultingData>();
        }
        return (List<ResultingData>)HttpContext.Current.Session["resultingData"];
    }
    int IncrementPollCount() {
        int pollCount = 3;
        if(HttpContext.Current.Session["pollCount"] != null) {
            pollCount = (int)HttpContext.Current.Session["pollCount"];
        }
        HttpContext.Current.Session["pollCount"] = ++pollCount;
        return pollCount;
    }
    XtraReport IWebDocumentViewerDrillThroughProcessor.CreateReport(DrillThroughContext context) {
        if(context.CustomData == "showResults") {
            context.AssignEditingFieldValuesToReport();
            IList<ResultingData> data = GetResultingData();
            ((PollingReport)context.Report).UpdateData(data);
            var resultingReport = new ResultingReport();
            resultingReport.ResultingReportDS.DataSource = data;
            resultingReport.Parameters["paramPollCount"].Value = IncrementPollCount();
            return resultingReport;
        } else {
            return new PollingReport();
        }
    }
}
